qcodes/instrument_drivers/Keysight/keysightb1500/KeysightB1500_base.py

```diff
@@ -282,3 +300,165 @@ def clear_buffer_of_error_message(self) -> None:
+        super().__init__(
+            name,
+            names=tuple(['param1', 'param2']),
+            units=tuple(['A', 'A']),
+            labels=tuple(['Param1 Current', 'Param2 Current']),
+            shapes=((1,),) * 2,
+            setpoint_names=(('Voltage',),) * 2,
+            setpoint_labels=(('Voltage',),) * 2,
+            setpoint_units=(('V',),) * 2,
+            **kwargs)
+        self._instrument = instrument
+        self.param1 = _FMTResponse(None, None, None, None)
+        self.param2 = _FMTResponse(None, None, None, None)
+        self.source_voltage = _FMTResponse(None, None, None, None)
+        self._fudge: float = 1.5
+
+    def get_raw(self):
+
+        measurement_mode = self._instrument.get_measurement_mode()
+        if len(measurement_mode['channels']) != 2:
+            raise ValueError('Two measurement channels are needed, one for '
+                             'gate current and other for source drain '
+                             'current.')
+
+        smu = self._instrument.by_channel[measurement_mode['channels'][0]]
+
+        if not smu.setup_fnc_already_run:
+            raise Exception(f'Sweep setup has not yet been run successfully on {smu.full_name}')
+
+        delay_time = smu.iv_sweep.step_delay()
+        if smu.average_coefficient < 0:
+            # negative coefficient means nplc and positive means just
+            # averaging
+            nplc = 128 * abs(self._instrument.average_coefficient)
+            power_line_time_period = 1 / smu.power_line_frequency
+            calculated_time = 2 * nplc * power_line_time_period
+        else:
+            calculated_time = smu.average_coefficient * \
+                              delay_time
+        num_steps = smu.iv_sweep.sweep_steps()
+        estimated_timeout = max(delay_time, calculated_time) * num_steps
+        new_timeout = estimated_timeout * self._fudge
+
+        format_and_mode = self._instrument.get_response_format_and_mode()
+        fmt_format = format_and_mode['format']
+        fmt_mode = format_and_mode['mode']
+        try:
+            self.root_instrument.write(MessageBuilder().fmt(1, 1).message)
+            with self.root_instrument.timeout.set_to(new_timeout):
+                raw_data = self._instrument.ask(MessageBuilder().xe().message)
+                parsed_data = parse_fmt_1_0_response(raw_data)
+        finally:
+            self.root_instrument.write(MessageBuilder().fmt(fmt_format,
+                                                            fmt_mode).message)
+
+        self.param1 = _FMTResponse(
+            *[parsed_data[i][::3] for i in range(0, 4)])
+        self.param2 = _FMTResponse(
+            *[parsed_data[i][1::3] for i in range(0, 4)])
+        self.source_voltage = _FMTResponse(
+            *[parsed_data[i][2::3] for i in range(0, 4)])
+
+        self.shapes = ((len(self.source_voltage.value),),) * 2
+        self.setpoints = ((self.source_voltage.value,),) * 2
```

Comment:
Does this actually work correctly. Previously updating the setpoints within the get_raw function was too late. But that may only be with the legacy loop?

Reply:
In the actual test, We got the correct results. We tested with many input points and it was working fine. So I think this is working fine.